### DISCUSSION TITLE
How to proper pin `pre-commit` back to 3.11

### DISCUSSION THREAD

**Discussion by Anselmoo** [Homebrew/discussions#4895] [Writing Formulae/Casks] [answered]
### Output of `brew config`

```shell
✗ brew config
HOMEBREW_VERSION: 4.1.19-35-g39c2296
ORIGIN: https://github.com/Homebrew/brew
HEAD: 39c2296ffbd9c2a225d6d64b234ef5584622599a
Last commit: 9 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: c20e982a3329ca27ebe98561d3ae1e7e6ac83c21
Core tap last commit: 14 hours ago
Core tap branch: master
Core tap JSON: 11 Nov 08:22 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.hCwYWxDCVz/org.xquartz:0
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 15.0.0 build 1500
Git: 2.42.1 => /opt/homebrew/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.5.2-arm64
CLT: 15.0.0.0.1.1694021235
Xcode: 15.0.1
Rosetta 2: false
```

### Output of `brew doctor`

```shell
➜ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  openssl@1.1
```

### Description of issue

> https://github.com/Homebrew/homebrew-core/blob/0a00c455796a30c1b06327b7fae5f630dcb0ac37/Formula/p/pre-commit.rb#L22C2-L22C27

Now, where `pre-commit` depends on `@python3.12`, the issue appears with `import distills` is not working. However, `mypy` depends on distills  to install additional dependencies:

```yaml
 - repo: https://github.com/pre-commit/mirrors-mypy
    rev: v1.6.1
    hooks:
      - id: mypy
        additional_dependencies:
          [
            types-toml==0.10.8.7,
            types-PyYAML==6.0.12.12,
            types-tabulate==0.9.0.3,
            types-mock==5.1.0.2,
            types-setuptools==68.2.0.1,
            numpy>=1.23.4,
            pydantic>=2.2,
            pytest==7.4.3,
          ]
```

As a result, the following error appears:

```terminal
      File "/opt/homebrew/Cellar/python@3.12/3.12.0/Frameworks/Python.framework/Versions/3.12/lib/python3.12/importlib/__init__.py", line 90, in import_module
        return _bootstrap._gcd_import(name[level:], package, level)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      File "<frozen importlib._bootstrap>", line 1381, in _gcd_import
      File "<frozen importlib._bootstrap>", line 1354, in _find_and_load
      File "<frozen importlib._bootstrap>", line 1304, in _find_and_load_unlocked
      File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
      File "<frozen importlib._bootstrap>", line 1381, in _gcd_import
      File "<frozen importlib._bootstrap>", line 1354, in _find_and_load
      File "<frozen importlib._bootstrap>", line 1325, in _find_and_load_unlocked
      File "<frozen importlib._bootstrap>", line 929, in _load_unlocked
      File "<frozen importlib._bootstrap_external>", line 994, in exec_module
      File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
      File "/private/var/folders/pb/1z5lx35d7y99w8rsvn2t6z8r0000gp/T/pip-build-env-5bnhlzgw/overlay/lib/python3.12/site-packages/setuptools/__init__.py", line 10, in <module>
        import distutils.core
    ModuleNotFoundError: No module named 'distutils'
Check the log at /Users/hahn/.cache/pre-commit/pre-commit.log
```

However, if using `pre-commit` from a local `virtualenv` with `python<3.12`, that's fine, so I am wondering if there any way to install it depending on `python<3.12`. Globally, setting up `pre-commit` with `3.11` would be great so other tools can have access to it to run the hooks.

thx for feedback

---

<img width="298" alt="image" src="https://github.com/Homebrew/discussions/assets/13209783/d9ac59f2-d8e5-4359-bef1-c83e5dc99055">

https://github.com/Homebrew/homebrew-core/blob/0a00c455796a30c1b06327b7fae5f630dcb0ac37/Formula/p/pre-commit.rb#L22C2-L22C27

**Comment by Anselmoo** [comment]
Dear @SMillerDev 

I am still looking for a bit of advice on how to install `pre-commit` using a pinned Python version `<3.12`. Although it works well for `black` or `ruff`, I have encountered issues when trying to incorporate other Python packages such as Numpy, due to the absence of `distutils.core`. 

I have been _unable_ to find an external installation for `distutils.core` for Python 3.12. However, **third-party packages** may be necessary for the extended notation in `MyPy`. 

Thank you very much for any kind of advice or forwarding this to an issue

**Comment by p-linnane** [comment, marked as answer]
We always use the latest version of dependencies, so you won't be able to do this and still use our binaries. You can edit the formula file for `pre-commit` and change the dependency to `python@3.11`. Then run `brew install -sv pre-commit` to build it from source. I'd caution you that this is unsupported, so you are on your own from here.
